### ISSUE TITLE
Precise scope of data queries for attributes only

### ISSUE THREAD

**Issue by dosse** [sdmx-twg/sdmx-rest#159]
When using the data query with `attributes=all` and `measures=none` with a partial filter, which attribute values should be returned?:  
1) Only the attribute values attached exactly at the partial key corresponding to the partial filter. Example: `AUS..Q`
1) The attribute values attached exactly at the partial key corresponding to the partial filter and the attribute values of all "lower" attachment levels (with more complete keys). Example: `AUS..Q`, `AUS.GDP.Q` and `AUS.PPP.Q`
1) The attribute values attached exactly at the partial key corresponding to the partial filter and the attribute values of all "higher" attachment levels (with less complete keys). Example: `AUS..Q`, `AUS..`, `..Q` and `..` (dataflow level).
1) All of the above. Example: `AUS..Q`, `AUS.GDP.Q`, `AUS.PPP.Q`, `AUS..`, `..Q` and `..` (dataflow level).

This also means, when using the data query with `attributes=all` and `measures=none` without a filter, should the attribute values attached at dataflow level only be returned, or the attribute values attached at any level (like for data)?  
Especially when looking at the referential metadata linked to a DSD, there are use cases where all 3 approaches would be needed. Would this require an additional parameter (option)?

**Comment by sosna** [comment]
Let's take the attached fictive DSD as example (and let's ignore OBS_COM):
![sc](https://user-images.githubusercontent.com/1846109/166638272-6ba7177f-6c0c-4fdf-8b2d-23db79ad21f4.png)

And let's take the following dataset:
![sc2](https://user-images.githubusercontent.com/1846109/166638412-063b03a1-95e2-4ed0-afce-e342240e5dcd.png)

If I query for `D.CHF.`, I would expect it to match the first row (leaving the header aside :-)). 

That means that all attributes applying to this row would be returned, regardless of their attachment level. That includes then the dataflow-level one `UNIT_MULT`, the group-level ones (`DECIMALS`, `UNIT`), the series-level one (`COLL`), and the observation-level one (`OBS_STATUS`).

So, in my understanding, the correct answer is **4**.

Does that clarify and, if so, do we need to do anything else for this issue?

**Comment by dosse** [comment]
Yes, thanks, @sosna, this clarifies well for the current expected behaviour.

Especially for ref. metadata linked to DSD, where the attachment is volatile if not pre-defined in the DSD, it would be great to have a new mean to filter by level, for example:
1) In our [Data Explorer](https://de-demo.siscc.org/vis?lc=en&df[ds]=staging%3ASIS-CC-stable&df[id]=MILLED_RICE&df[ag]=TN1&df[vs]=1.0&av=true&pd=2015%2C2018&dq=..A&vw=tb), a user can see the content of a ref. metadata value at any level (by clicking on the blue `(i)` icons, and we display at the same time the ref. metadata attached at all higher levels. For that we introduced a specific HTTP header `x-level: upperOnly` 
2) In our DataLifeCycleManager, we allow getting (and then editing) specific ref. metadata values related to a dimension filter. It would be great if the user could fine-tune the level, to get and edit only the needed ones, e.g. get all ref. metadata values attached exactly at partial key `D.CHF.`

Such a filtering feature seems quite useful, what do you think?

**Comment by sosna** [comment]
Thanks, @dosse 

The reference metadata is the part of the spec I know the least, so I would prefer to forward the question to @agent96 :-). It definitely looks useful but, as mentioned, I don't know enough about what's possible there.

**Comment by agent96** [comment]
In @dosse example it read this to mean that the client has queried for the data but excluded the reference metadata in the response - as such it needs mechanisms to allow the user to get the metadata on demand.  

There are 2 requirements listed, which I have elaborated below on points 2 and 3, but it would seem to me that point 1 is critical to make this work?  If not how would the client know where to put the 'more info' indicators.

1. The client needs to know there are metadata attributes reported against certain keys/partial keys without the client actually having the metadata - thus enabling the client to know that it can indicate there is more detail - currently I do not think this is the case
2. The client needs to be able to get reported metadata attributes with no observations for a specific key or partial key and have it include any metadata which matches part of the key with other parts wildcarded (i.e. client queries for metadata against key D.CHF. and they get metadata attached to this partial key and attached to just .CHF. and just D.. but not D.GBP) - question, do they get metadata for D.CHF.EMP?
3. The client needs to be able to get reported metadata attributes which exactly match the key D.CHF - in example #2 this would only return 1 item, that against D.CHF and not D.CHF.EMP and not D.. or .CHF.

SDMX 3.0 design of the metadata against data such that it acts like data attributes, with some subtle exception on how it can be reported.  As such it would seem to me that the best way to solve 2 and 3 would be to create a solution which applies to any attribute (metadata or standard).  I think some thought needs to go into point 1.

**Comment by dosse** [comment]
Thanks @sosna and @agent96 !

On @agent96's 3 points:

1. "_know there are metadata attributes reported against certain keys/partial keys without the client actually having the metadata_"  
   Since we do not have yet any proper SDMX feature to get the ref metadata availability related to a filter, we have solved this by automatically generating annotations "HAS_METADATA" attached to the dimension combinations exposed in the SDMX-JSON data message obtained with `attributes=dsd&measures=none` in the new `dimensionGroupAttributes` object, e.g.
   
   ```json
    "data": {
        "dataSets": [{
            ...,
            "dimensionGroupAttributes": {
                ":::": [3],
                "0:::": [3],
                "0::0:": [3],
                "1:0:0:0": [3],
                "1:0:0:1": [3],
                "2:0:0:0": [3],
                "2:0:0:": [3]
            },
            ...
        }],
        "structures": [{
            ...,
            "attributes": {
               ...,
                "dimensionGroup": [],
                ...
            },
            "annotations": [
                ...,
                {"type": "HAS_METADATA"}
            ],
            ...
        }]
    }
   ```
   Thus 1 is ok for us.
2. "_be able to get reported metadata attributes with no observations for a specific key or partial key and have it include any metadata which matches part of the key with other parts wildcarded_"  
   We are using the URL `/*.TONES.A?attributes=msd&measures=none` to get (only) the reported metadata attributes for a specific key or partial key. And here comes the background reason for this GitLab issue here. The client needs to be able indicating the "scope" of the filter, e.g. to get the exact match plus all higher-level ref metadata, or to get only ref. metadata at the specified partial key (see 3. below).  We have solved this so far with a non-SDMX HTTP header **`x-level: upperOnly`**. By default, when this header is not set, we return all levels - see above confirmation from @sosna about the appropriateness of returning all levels by default.
3. "_be able to get reported metadata attributes which exactly match the (partial) key_"  
   Similar to 2., we have solved this so far with a non-SDMX HTTP header **`x-level=currentOnly`**.

Thus all is working for us, except that we currently use a non-SDMX HTTP header. It would be great if this could be added to the standard.
